test/functional/test-batched-xhr.js

```diff
@@ -14,21 +14,12 @@
  * limitations under the License.
  */
 
-import {FetchResponse, fetchPolyfill} from '../../src/service/xhr-impl';
+import {FetchResponse} from '../../src/service/xhr-impl';
 import {batchedXhrServiceForTesting} from '../../src/service/batched-xhr-impl';
-
+import {installDocService} from '../../src/service/ampdoc-impl';
```

Comment:
We do this a lot already but... instead of introducing this dependency to each test, can we stub this with a fake object instead? This will help keep unit tests more hermetic.

```js
import {Services} from '../../src/services';

beforeEach(() => {
  env.sandbox.stub(Services, 'ampdocServiceFor').returns({isSingleDoc: () => false});
});
```

Reply:
Thanks, done.

For test-xhr.js, the XHR service is currently eagerly created before all tests are executed. I had to lazily create it as the dependency stub isn't set up until the sandbox is created in `beforeEach`.

Do we usually stub/mock all the dependencies of the object under test, or just bulky dependencies such as services? In other projects I work on, I tend to follow the practice to prefer using real objects for dependencies so long as the real object is fast and deterministic. The arguments are:

1. The "unit" in a unit test isn't necessarily a single class/file but rather a complete code path of a single public API.
2. If there's a bug in the dependency, we want the current test to also fail as that mirrors the fact that the current system under test will also fail in production.
3. Using real objects allow testing "states" rather than interactions.

Classicists vs Mockist. I guess this varies from projects to projects. :)

Reply:
Interesting point, but perhaps this is just semantics. IMO ideally you'd have tests for both: "unit" and "integration" tests (for their traditional definitions). 

One issue is I can see with that proposal is a bug in a dependency could cause a code path in the current class to not be tested (e.g. if `isSingleDoc()` always returned true).